kong/runloop/handler.lua

```diff
@@ -721,25 +724,114 @@ return {
 
       local trusted_ip = kong.ip.is_trusted(realip_remote_addr)
       if trusted_ip then
-        forwarded_proto = var.http_x_forwarded_proto or var.scheme
-        forwarded_host  = var.http_x_forwarded_host  or var.host
-        forwarded_port  = var.http_x_forwarded_port  or var.server_port
+        forwarded_proto = var.http_x_forwarded_proto or scheme
+        forwarded_host  = var.http_x_forwarded_host  or host
+        forwarded_port  = var.http_x_forwarded_port  or port
 
       else
-        forwarded_proto = var.scheme
-        forwarded_host  = var.host
-        forwarded_port  = var.server_port
+        forwarded_proto = scheme
+        forwarded_host  = host
+        forwarded_port  = port
       end
 
-      local protocols = route.protocols
-      if (protocols and
-          protocols.https and not protocols.http and forwarded_proto ~= "https")
+      local protocols = route and route.protocols
+      if (protocols and protocols.https and not protocols.http and
+          forwarded_proto ~= "https")
       then
         ngx.header["connection"] = "Upgrade"
         ngx.header["upgrade"]    = "TLS/1.2, HTTP/1.1"
         return kong.response.exit(426, { message = "Please use HTTPS protocol" })
       end
 
+      if not service then
+        -----------------------------------------------------------------------
+        -- Serviceless HTTP / HTTPS / HTTP2 route
+        -----------------------------------------------------------------------
+        local service_scheme
+        local service_host
+        local service_port
+
+        -- 1. try to find information from a request-line
+        if var.http2 == "" then
+          local request_line = var.request
+          if request_line then
+            local matches = ngx.re.match(request_line, [[\w+ (\S+)]], "ajos")
+            if matches and matches[1] then
+              local uri = url.parse(matches[1])
```

Comment:
Could this be done with a regex that extracts matches?

I think this could be a hot path. But as it's new functionality we could just fix it later.
maybe add a `-- TODO: optimize`?

Reply:
I rewrote that without `socket.url`.